akka-stream/src/main/scala/akka/stream/scaladsl/Queue.scala

```diff
@@ -32,6 +32,23 @@ trait SourceQueue[T] {
 }
 
 /**
+ * This trait adds completion support to [[SourceQueue]].
+ */
+trait SourceQueueWithComplete[T] extends SourceQueue[T] {
+  /**
+   * Complete the stream normally. Use `watchCompletion` to be notified of this
+   * operation’s success.
+   */
+  def complete(): Unit
+
+  /**
+   * Complete the stream with a failure. Use `watchCompletion` to be notified of this
+   * operation’s success.
+   */
+  def fail(ex: Throwable): Unit
```

Comment:
>not sure what you are getting at here: there is no feedback, so you can call it any number of times to get back any number of Unit values—is that not idempotent enough?

My point is that if you *Accidentally* or as a result of a race-condition call fail after complete, or fail after fail, or complete after complete, you have no indicator that you have a logic bug?